Flaky tests with TestCustomResolver and TestOpenTelemetryTracing

For `TestCustomResolver`, CI is sometimes returning:

```
=== RUN   TestCustomResolver
    grpc_dialer_test.go:206: 
        	Error Trace:	grpc_dialer_test.go:206
        	Error:      	Not equal: 
        	            	expected: 1
        	            	actual  : 2
        	Test:       	TestCustomResolver
--- FAIL: TestCustomResolver (0.12s)
```

And for `TestOpenTelemetryTracing`, CI is sometimes returning:
```
=== CONT  TestIntegrationSuite
    integration_test.go:1610: 
        	Error Trace:	integration_test.go:1610
        	            				integration_test.go:1497
        	Error:      	Not equal: 
        	            	expected: &interceptortest.SpanInfo{Name:"root-span", Children:[]*interceptortest.SpanInfo{(*interceptortest.SpanInfo)(0xc0008a8930), (*interceptortest.SpanInfo)(0xc0008a8990), (*interceptortest.SpanInfo)(0xc0008a89f0)}}
        	            	actual  : &interceptortest.SpanInfo{Name:"root-span", Children:[]*interceptortest.SpanInfo{(*interceptortest.SpanInfo)(0xc0008a8330), (*interceptortest.SpanInfo)(0xc0008a8390), (*interceptortest.SpanInfo)(0xc0008a83f0)}}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -12,3 +12,3 @@
        	            	      Name: (string) (len=29) "RunWorkflow:SignalsAndQueries",
        	            	-     Children: ([]*interceptortest.SpanInfo) (len=4) {
        	            	+     Children: ([]*interceptortest.SpanInfo) (len=5) {
        	            	       (*interceptortest.SpanInfo)({
        	            	@@ -25,2 +25,6 @@
        	            	        }
        	            	+      }),
        	            	+      (*interceptortest.SpanInfo)({
        	            	+       Name: (string) (len=32) "SignalChildWorkflow:start-signal",
        	            	+       Children: ([]*interceptortest.SpanInfo) <nil>
        	            	       }),
        	Test:       	TestIntegrationSuite
=== CONT  TestIntegrationSuite/TestOpenTelemetryTracing
```

Need to fix.

--- a/internal/grpc_dialer_test.go
+++ b/internal/grpc_dialer_test.go
@@ -27,6 +27,7 @@ package internal
 import (
 	"context"
 	"errors"
+	"fmt"
 	"log"
 	"net"
 	"strings"
@@ -179,7 +180,7 @@ func TestDialOptions(t *testing.T) {
 }
 
 func TestCustomResolver(t *testing.T) {
-	ctx, cancel := context.WithCancel(context.Background())
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
 	defer cancel()
 	// Create two gRPC servers
 	s1, err := startAdditionalHostPortsGRPCServer()
@@ -188,8 +189,6 @@ func TestCustomResolver(t *testing.T) {
 	s2, err := startAdditionalHostPortsGRPCServer()
 	require.NoError(t, err)
 	defer s2.Stop()
-	// Wait a bit to ensure they are serving
-	time.Sleep(100 * time.Millisecond)
 
 	// Register resolver for both IPs and create client using it
 	scheme := "test-resolve-" + uuid.New()
@@ -242,7 +241,34 @@ func startAdditionalHostPortsGRPCServer() (*customResolverGRPCServer, error) {
 			log.Fatal(err)
 		}
 	}()
-	return s, nil
+
+	// Wait until health reports serving
+	return s, s.waitUntilServing()
+}
+
+func (c *customResolverGRPCServer) waitUntilServing() error {
+	// Try 20 times, waiting 100ms between
+	var lastErr error
+	for i := 0; i < 20; i++ {
+		conn, err := grpc.Dial(c.addr, grpc.WithInsecure())
+		if err != nil {
+			lastErr = err
+		} else {
+			resp, err := grpc_health_v1.NewHealthClient(conn).Check(context.Background(), &grpc_health_v1.HealthCheckRequest{
+				Service: healthCheckServiceName,
+			})
+			_ = conn.Close()
+			if err != nil {
+				lastErr = err
+			} else if resp.Status != grpc_health_v1.HealthCheckResponse_SERVING {
+				lastErr = fmt.Errorf("last status: %v", resp.Status)
+			} else {
+				return nil
+			}
+		}
+		time.Sleep(100 * time.Millisecond)
+	}
+	return fmt.Errorf("failed waiting, last error: %w", lastErr)
 }
 
 func (c *customResolverGRPCServer) SignalWorkflowExecution(

--- a/test/integration_test.go
+++ b/test/integration_test.go
@@ -1614,21 +1614,22 @@ func (ts *IntegrationTestSuite) openTelemetrySpanChildren(
 	spans []sdktrace.ReadOnlySpan,
 	parentID trace.SpanID,
 ) (ret []*interceptortest.SpanInfo) {
-	var lastSpan *interceptortest.SpanInfo
+	seenSpans := map[string]*interceptortest.SpanInfo{}
 	for _, s := range spans {
 		if s.Parent().SpanID() == parentID {
 			// In cases where we have disabled the cache, the same interceptors get
 			// called many times with replayed values which create replayed spans. We
-			// can't disable  spans during replay because they are sometimes the ones
-			// that code paths continue on so they need spans. So in this case we will
-			// reuse adjacent spans of the same name.
-			children := ts.openTelemetrySpanChildren(spans, s.SpanContext().SpanID())
-			if ts.config.maxWorkflowCacheSize == 0 && lastSpan != nil && lastSpan.Name == s.Name() {
-				lastSpan.Children = append(lastSpan.Children, children...)
-			} else {
-				lastSpan = interceptortest.Span(s.Name(), children...)
-				ret = append(ret, lastSpan)
+			// can't disable spans during replay because they are sometimes the ones
+			// that code paths continue on so they need spans. Since the tests don't
+			// expect any spans of the same name at the same level, we will just reuse
+			// spans we find with the same name to prevent this.
+			span := seenSpans[s.Name()]
+			if span == nil {
+				span = interceptortest.Span(s.Name())
+				ret = append(ret, span)
+				seenSpans[s.Name()] = span
 			}
+			span.Children = append(span.Children, ts.openTelemetrySpanChildren(spans, s.SpanContext().SpanID())...)
 		}
 	}
 	return

